File: approaches/mcmc/poisson-gamma.py
import pymc3 as pm
import numpy as np
from tabulate import tabulate
from scipy.optimize import linprog
import scipy.stats as stats 
import matplotlib
from matplotlib import pyplot as plt
import matplotlib.animation as animation

d0 = [20, 28, 24, 20, 23]                # observed demand samples

with pm.Model() as m:
    d = pm.Gamma('theta', 1, 1)          # prior distribution
    pm.Poisson('d0', d, observed = d0)   # likelihood
    samples = pm.sample(10000)           # draw samples from the posterior

    
seaborn.distplot(samples.get_values('theta'), fit=scipy.stats.gamma)

p = np.linspace(10, 16)   # price range
d_means = np.exp(s.log_b + s.a * np.log(p).reshape(-1, 1))

plt.plot(p, d_means, c = 'k', alpha = 0.01)
plt.plot(p0, d0, 'o', c = 'r')
plt.show()

File: approaches/mcmc/poisson-gamma.jl
using StanSample, MCMCChains, StatsPlots

ProjDir = @__DIR__

#=
with pm.Model() as m:
    d = pm.Gamma('theta', 1, 1)          # prior distribution
    pm.Poisson('d0', d, observed = d0)   # likelihood
    samples = pm.sample(10000)           # draw samples from the posterior
=#

demand_model = "
data {
  int<lower=1> N;
  int<lower=0> d[N];
}
parameters {
  real<lower=0> theta;
}
model {
  theta ~ gamma(1,1);
  d ~ poisson(theta);
}
";

d = [20, 28, 24, 20, 23]        # observed demand samples
demand_data = Dict(
  :N => length(d),
  :d => d
)

# Keep tmpdir across multiple runs to prevent re-compilation
tmpdir = joinpath(@__DIR__, "tmp")

sm = SampleModel("demand-price-model", demand_model)

rc = stan_sample(sm; data=demand_data)

if success(rc)
  # Convert to an MCMCChains.Chains object
  chns = read_samples(sm; output_format=:mcmcchains)
  show(chns)
  plot(chns)
  savefig(joinpath(@__DIR__, "fig_01.png"))
end
